pkg/migration-tool/writer/shards.go

```diff
@@ -0,0 +1,189 @@
+package writer
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/gogo/protobuf/proto"
+	"github.com/golang/snappy"
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/prometheus/prompb"
+	"github.com/prometheus/prometheus/storage/remote"
+	"github.com/timescale/promscale/pkg/log"
+	"github.com/timescale/promscale/pkg/migration-tool/utils"
+)
+
+// shardsSet represents a set of shards. It consists of configuration related
+// to handling and management of shards.
+type shardsSet struct {
+	num         int
+	set         []*shard
+	cancelFuncs []context.CancelFunc
+
+	done    chan struct{}
+	errChan chan error
+}
+
+func newShardsSet(writerCtx context.Context, writeURL string, capacity int) (*shardsSet, error) {
+	var (
+		set         = make([]*shard, capacity)
+		cancelFuncs = make([]context.CancelFunc, capacity)
+	)
+	ss := &shardsSet{
+		num:     capacity,
+		done:    make(chan struct{}, capacity),
+		errChan: make(chan error, capacity),
+	}
+	for i := 0; i < capacity; i++ {
+		client, err := utils.NewClient(fmt.Sprintf("writer-shard-%d", i), writeURL, utils.Write, model.Duration(defaultWriteTimeout))
+		if err != nil {
+			return nil, fmt.Errorf("creating write-shard-client-%d: %w", i, err)
+		}
+		ctx, cancelFunc := context.WithCancel(writerCtx)
+		shard := &shard{
+			ctx:          ctx,
+			client:       client,
+			queue:        make(chan *[]prompb.TimeSeries),
+			copyShardSet: ss,
+		}
+		cancelFuncs[i] = cancelFunc
+		set[i] = shard
+	}
+	ss.set = set
+	ss.cancelFuncs = cancelFuncs
+	// Run the shards.
+	for i := 0; i < ss.num; i++ {
+		go ss.set[i].run(i)
+	}
+	return ss, nil
+}
+
+// scheduleTS batches the time-series based on the available shards. These batches are then
+// fed to the shards in a single go which then consume and push to the respective clients.
+func (s *shardsSet) scheduleTS(ts []prompb.TimeSeries) {
+	var (
+		tsEachShard = len(ts) / s.num
+		batches     = make([]*[]prompb.TimeSeries, s.num)
+		tsIndex     = 0
+	)
+	for i := tsIndex; i < s.num; i++ {
+		var batch []prompb.TimeSeries
+		if i == s.num-1 {
+			batch = ts[tsIndex:]
+		} else {
+			batch = ts[tsIndex : tsIndex+tsEachShard]
+			tsIndex += tsEachShard
+		}
+		batches[i] = &batch
+	}
```

Comment:
I think we can avoid the allocation of the batches slice. As you create a batch in the cycle just enqueue it at s.set[tsIndex].queue and no need to call assign. Or I'm missing something

Reply:
Oh yes. Thank you for the advice. That will help save a lot of very useful memory.